client/client-lib/src/lib.rs

```diff
@@ -430,23 +431,51 @@ impl<T: AsRef<ClientConfig> + Clone> Client<T> {
 }
 
 impl Client<UserClientConfig> {
-    pub async fn fetch_gateway(&self) -> Result<LightningGateway> {
-        // fetch gateway from db
-        if let Some(gateway) = self
-            .context
+    pub async fn fetch_registered_gateways(&self) -> Result<Vec<LightningGateway>> {
+        let gateways = self.context.api.fetch_gateways().await?;
+        let current_active = self.fetch_active_gateway().await;
+
+        if current_active.is_some() {
+            // We have an active gateway in this client. We should update the list of registered gateways with that information.
+            Ok(gateways
+                .into_iter()
+                .map(|mut gateway| {
+                    if gateway.node_pub_key == current_active.clone().unwrap().node_pub_key {
+                        gateway.active = true;
+                    } else {
+                        gateway.active = false;
+                    }
+                    gateway
+                })
+                .collect())
+        } else {
+            Ok(gateways)
+        }
+    }
+    pub async fn fetch_active_gateway(&self) -> Option<LightningGateway> {
+        self.context
             .db
             .get_value(&LightningGatewayKey)
             .expect("DB error")
-        {
-            return Ok(gateway);
-        }
-
-        // if db is empty, fetch from federation and save to db
-        let gateways = self.context.api.fetch_gateways().await?;
+    }
+    pub async fn switch_active_gateway(
+        &self,
+        node_pub_key: Option<secp256k1::PublicKey>,
+    ) -> Result<LightningGateway> {
+        let gateways = self.fetch_registered_gateways().await?;
         if gateways.is_empty() {
             return Err(ClientError::NoGateways);
         };
-        let gateway = gateways[0].clone();
+        let mut gateway = match node_pub_key {
+            // If a pubkey was provided, try to select and activate a gateway with that pubkey.
+            Some(pub_key) => gateways
+                .into_iter()
+                .find(|g| g.node_pub_key == pub_key)
+                .ok_or(ClientError::GatewayNotFound)?,
+            // Otherwise (no pubkey provided), select and activate the first registered gateway.
+            None => gateways[0].clone(),
```

Comment:
I don't think the behavior should be like this. You could do something like this on the Command struct (in mint-client-cli)
```
#[clap(arg_required_else_help = true)]
ListGateways { }
```

Reply:
Just saw this would never really been reached anyway..Why put an `Option<>` around the PublicKey in the first place ?

Reply:
Saw the usecase of the Option after the comment above, imo should still not be done like this